oauth2_provider/models.py

```diff
@@ -433,6 +485,105 @@ class Meta(AbstractRefreshToken.Meta):
         swappable = "OAUTH2_PROVIDER_REFRESH_TOKEN_MODEL"
 
 
+class AbstractIDToken(models.Model):
+    """
+    An IDToken instance represents the actual token to
+    access user's resources, as in :openid:`2`.
+
+    Fields:
+
+    * :attr:`user` The Django user representing resources' owner
+    * :attr:`token` ID token
+    * :attr:`application` Application instance
+    * :attr:`expires` Date and time of token expiration, in DateTime format
+    * :attr:`scope` Allowed scopes
+    """
+
+    id = models.BigAutoField(primary_key=True)
+    user = models.ForeignKey(
+        settings.AUTH_USER_MODEL,
+        on_delete=models.CASCADE,
+        blank=True,
+        null=True,
+        related_name="%(app_label)s_%(class)s",
+    )
+    token = models.TextField(unique=True)
```

Comment:
This is a bit of a pain - only noticed this as we started to move our own app in to QA with Google Cloud SQL - mysql doesn't support indexes on TextField, it only allows a partial index of the text. Its my fault, I only tested against postgres, sqlite and mariadb in development - mariadb supports indexing `TEXT` fields by hashing the text and indexing that instead.

Even if we decided to add a partial index for mysql, its a real pain to do so with django - you have to add it as a RunSQL migration and tell it that its a partial index. I'm investigating to see whether there is a way to do this operation just on mysql DBs. It would also mean that there is a possibility that the uniqueness constraint would be violated if just the first 255 characters are the same (the limit on a UTF-8 string index in mysql 5.7/5.8). Probably would have to be a non-unique index in that case.

If we can't, we'd have to ship instructions on how to swap the model so that mysql users can create a migration themselves that will work and give instructions for that.

A final, more drastic action would be to not store the ID tokens at all. Currently, we only retrieve ID tokens to validate that an ID token that we've received is valid. It is possible instead to validate them by verifying the JWT token signature, and checking `iss`, `nbf`, and `exp`. Of course, this makes it impossible to revoke an ID token when revoking an access token.

Reply:
As an example of how horrible this is, this is how we worked around it for now on our QA site:

```python


# Extend the IDToken model so that we can manually tweak the migration

class IDToken(AbstractIDToken):
    """
    We have to override this model because mysql cannot handle indexes on a
    TextField, and the standard migrations will break when run on mysql.

    By overriding the field and marking it as swapped for this modedl, Django
    won't attempt to run the migration for the base model and we can use our
    hand crafted migration to create the table as best we can.
    """

    pass


# Create a custom DB migration operation to do things differently on mysql

class MysqlOnlyRunSQL(migrations.RunSQL):
    """
    Mysql cannot create indexes on a whole text field, and django will not auto
    restrain the index to what mysql does support.

    We can hack this in a way that adds what mysql can support with a RunSQL
    migration, but that will fail on non mysql databases. This class wraps RunSQL
    and if the migration is run on a non mysql database, it will instead run the
    operations specified in the state_operations.

    This allows django to think that the desired state has been achieved on both
    mysql and non-mysql databases.
    """

    @property
    def reversible(self):
        """
        Can only be reversed if both the RunSQL and regular operations can be reversed
        """
        return self.reverse_sql is not None and all(
            op.reversible for op in self.state_operations
        )

    def database_forwards(self, app_label, schema_editor, from_state, to_state):
        """
        If we are using a mysql schema editor, run the supplied SQL.
        Otherwise, apply the operations in state_operations
        """
        if "mysql" in schema_editor.__class__.__module__:
            super().database_forwards(app_label, schema_editor, from_state, to_state)
        else:
            for op in self.state_operations:
                op.database_forwards(app_label, schema_editor, from_state, to_state)

    def database_backwards(self, app_label, schema_editor, from_state, to_state):
        """
        If we are using a mysql schema editor, run the supplied SQL.
        Otherwise, apply the operations in state_operations
        """
        if "mysql" in schema_editor.__class__.__module__:
            super().database_backwards(app_label, schema_editor, from_state, to_state)
        else:
            for op in self.state_operations:
                op.database_backwards(app_label, schema_editor, from_state, to_state)


# Handcraft the initial migration

        migrations.CreateModel(
            name="IDToken",
            fields=[
                ("id", models.BigAutoField(primary_key=True, serialize=False)),
                ("token", models.TextField()),
                ("expires", models.DateTimeField()),
                ("scope", models.TextField(blank=True)),
                ("created", models.DateTimeField(auto_now_add=True)),
                ("updated", models.DateTimeField(auto_now=True)),
                (
                    "application",
                    models.ForeignKey(
                        blank=True,
                        null=True,
                        on_delete=django.db.models.deletion.CASCADE,
                        to=oauth2_settings.APPLICATION_MODEL,
                    ),
                ),
                (
                    "user",
                    models.ForeignKey(
                        blank=True,
                        null=True,
                        on_delete=django.db.models.deletion.CASCADE,
                        related_name="pots_oauth_idtoken",
                        to=settings.AUTH_USER_MODEL,
                    ),
                ),
            ],
            options={
                "abstract": False,
            },
        ),
        # Hack to add partial indexes only on mysql, and full indexes on other DBs
        MysqlOnlyRunSQL(
            "CREATE INDEX `token_idx` ON `pots_oauth_idtoken` (`token`(255))",
            reverse_sql="ALTER TABLE `pots_oauth_idtoken` DROP INDEX `token_idx`",
            state_operations=[
                migrations.AlterField(
                    model_name="IDToken",
                    name="token",
                    field=models.TextField(unique=True),
                ),
            ],
        ),
```

Big eurgh!

Reply:
@tevansuk Have you considered splitting the JWT into its 3 component parts in the model to enable the uniqueness constraint? The SHA-256 is guaranteed to be 64 bytes, so easily uniquely indexed. You don't really need to uniquely constrain the _entire_ JWT as the SHA-256 signature is sufficient to establish uniqueness since it is a secure hash of the header and payload.

```python
    header = models.TextField()
    payload = models.TextField()
    signature = models.CharField(unique=True, max_length=64)
```

I haven't yet read the code to see how `AbstractIDToken` is used so I may be totally off-base here.

Reply:
This seems like a sensible approach, I'll give this a try and see what will need to be changed. Thanks Alan :+1:

Reply:
I think using the signature will also not work - a SHA-256 signature using a 2048 bit key will be 2048 bits/256 bytes long in binary, or 342 bytes long when base64 urlencoded.

I think the best way forward will be to:

* include a `jti` (`uuid.uuid4()`) in the id token
* store the `jti` and the token separately in the DB, indexing the `jti` field
* use the `jti` as the unique key for the token

When we receive a token and want to retrieve it from the DB:

* verify the token's signature
* extract the `jti`
* look up the `IDToken` in the database using the `jti`

Whilst we could possibly index the SHA-256 signature as binary for a 2048 bit key, we might hit DB concerns again for a larger key, or if we later wanted to use a different signature algorithm. Indexing the `jti` claim would work whatever hash algorithm was used.

Reply:
Hmm, according to [this mysql 5.7 doc](https://dev.mysql.com/doc/refman/5.7/en/innodb-limits.html): 

    The index key prefix length limit is 767 bytes for InnoDB tables that use the REDUNDANT or COMPACT row format. For 
    example, you might hit this limit with a column prefix index of more than 255 characters on a TEXT or VARCHAR 
    column, assuming a utf8mb3 character set and the maximum of 3 bytes for each character.

So that limit is for 255 **multi-byte** characters. Wouldn't this be a maximum of 256 single-byte characters we are talking about?

Reply:
Yes, 256 binary characters, which we wouldn't want to stuff in to a django `CharField`! 

Part of the problem is Django sets its own limits when dealing with mysql; it has no idea (and no way to specify) what the content you are storing in a CharField is, and so if you are using mysql [it sets the limit](https://docs.djangoproject.com/en/2.2/ref/databases/#notes-on-specific-fields) on the maximum length of a `CharField` to be 255, and doesn't support db indexes on `TextField`. Django only provides one field for storing binary data, `BinaryField`, which on mysql maps to `LONGBLOB`; this suffers from the same issue that we have with `TEXT`, mysql requires a maximum key length for an index, and Django provides no way for us to specify this.

`jti` is designed for specifically for identifying a JWT; we have a JWT that is signed by us, we've verified the signature, `exp` and `nbf`, we now just want to know if it is still valid in the database.

I'm not even sure we should be storing the contents of the ID token. We use the ID token model for:

* Checking to see if the token exists in our database and hasn't been revoked
* Determining the application it was issued to
* Determining the scopes it was issued for

None of this requires storing the actual token itself if we are indexing by the `jti`. If access to the claims is also required; well we've received the JWT and validated it, so the claims are available directly in the JWT.

This is also desirable from a PII point of view; if we don't store the contents of the ID token, then there is no PII at all in the `IDToken` table.

Reply:
Looks like you resolved this with d6f6082. Sorry for my delay in getting back.